### ISSUE TITLE
Possible memory leak in nn.ConvTranspose2d

### ISSUE THREAD

**Issue by emitch** [pytorch/pytorch#7248]
## Problem Description

When training a 2d UNet, memory usage steadily increases each batch until memory is exhausted. The issue seems to be related to the nn.ConvTranspose2d module, which is used in the upsampling part of the UNet to increase feature map resolution. Simply replacing each nn.ConvTranspose2d with a nn.Upsample eliminates the leak.

My network is defined as follows:

    import torch
    import torch.nn as nn
    import torch.nn.functional as F
    from torch.autograd import Variable
    
    class Conv(nn.Module):
        def __init__(self, in_ch, out_ch, f=nn.ReLU(inplace=True), k=3, same=True, f_out=None, groups=1):
            super(Conv, self).__init__()
            self.f = f
            self.f_out = f_out
            self.conv1 = nn.Conv2d(in_ch, out_ch, k, padding=(k-1)/2 if same else 0, groups=groups)
            self.conv2 = nn.Conv2d(out_ch, out_ch, k, padding=(k-1)/2 if same else 0)
    
        def forward(self, x):
            out = self.f(self.conv1(x))
            out = self.f(self.conv2(out)) if self.f_out is None else self.f_out(self.conv2(out))
            return out
    
    class Down(nn.Module):
        def __init__(self, in_ch, out_ch, f=nn.ReLU(inplace=True), k=3, same=True):
            super(Down, self).__init__()
            self.f = f
            self.downsample = nn.MaxPool2d(2)
            self.conv1 = nn.Conv2d(in_ch, in_ch, k, padding=(k-1)/2 if same else 0)
            self.conv2 = nn.Conv2d(in_ch, out_ch, k, padding=(k-1)/2 if same else 0)
    
        def forward(self, x):
            out = self.f(self.downsample(x))
            out = self.f(self.conv1(out))
            return self.f(self.conv2(out))
    
    
    class Up(nn.Module):
        def __init__(self, in_ch, out_ch, f=nn.ReLU(inplace=True), k=3, same=True, cat=True):
            super(Up, self).__init__()
            self.f = f
            #self.upsample = nn.ConvTranspose2d(in_ch, in_ch, k, stride=2, padding=(k-1)/2 if same else 0, output_padding=1)                                                                                                                                                                                                                                                  
            self.upsample = nn.Upsample(scale_factor=2, mode='bilinear')
            self.conv1 = nn.Conv2d(in_ch + out_ch if cat else in_ch, out_ch, k, padding=(k-1)/2 if same else 0)
            self.conv2 = nn.Conv2d(out_ch, out_ch, k, padding=(k-1)/2 if same else 0)
    
        def forward(self, x, y=None):
            out = self.f(self.upsample(x))
            if y is not None:
                out = torch.cat((out, y),1)
            out = self.f(self.conv1(out))
            return self.f(self.conv2(out))
    
    class UNet(nn.Module):
        def __init__(self, in_ch=2, out_ch=2, depth=4, fm=24, k=3, same=True):
            super(UNet, self).__init__()
            
            print('--------- Building UNet with depth ' + str(depth) + ' and kernel size ' + str(k) + ' ---------')
            
            self.downModules = nn.ModuleList([Down(fm, fm, k=k, same=same) for i in range(depth)])
            self.upModules = nn.ModuleList([Up(fm, fm, k=k, same=same) for i in range(depth)])
            
            self.embed_in = Conv(in_ch, fm, k=7, same=same, groups=2)
            self.embed_out = Conv(fm, out_ch, same=same, f_out=lambda x: x)
            self.f = nn.ReLU()
    
        def forward(self, x):
            out = self.embed_in(x)
            downPass = [out]
            for down in self.downModules:
                downPass.append(down(downPass[-1]))
            upPass = [downPass[-1]]
            for idx, up in enumerate(self.upModules):
                upPass.append(up(upPass[-1], downPass[-(idx+2)]))
            out = self.embed_out(upPass[-1])
            return out.permute(0,2,3,1)

Commenting out the nn.Upsample() module and uncommenting the nn.ConvTranspose2d() module in the Up module triggers the issue.

## System Info

PyTorch version: 0.3.0.post4
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 2.7
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 390.25
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/lib64/libcudnn.so
/usr/local/cuda-8.0/lib64/libcudnn.so.7
/usr/local/cuda-8.0/lib64/libcudnn.so.7.0.5
/usr/local/cuda-8.0/lib64/libcudnn_static.a
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudnn-900fef33.so.7.0.5
/usr/local/lib/python3.5/dist-packages/torch/lib/libcudnn-a2b758a6.so.7.0.3

Versions of relevant libraries:
[pip] numpy (1.14.2)
[pip] torch (0.3.0.post4)
[pip] torchfile (0.1.0)
[pip] torchnet (0.0.1)
[pip] torchvision (0.2.0)
[conda] Could not collect

**Comment by li-roy** [comment]
I'm having a bit of trouble reproducing with what you've given. Could you provide some code that reproduces the problem?

**Comment by weiyangfb** [comment]
@emitch has this been resolved?

**Comment by emitch** [comment]
Unfortunately I no longer have access to the machine that was seeing this behavior. On a new machine, I'm not seeing the same problem (same PyTorch version). Closing for now until someone else observes the issue.
